### PR TITLE
chore(client): retry api version discovery for at least 30 seconds

### PR DESCRIPTION
**Pull request by dpc**
During debugging with joschisan, we've spotted a case where in a test the Federation was not yet fully up (peer api servers were down), yet the test Client was already attempting to join with a pre-prepared config.

Normally in this case fetching the config would fail, but that was passed straight from the server fixture, so api discovery call was actually the first call. It queried all peers, all were down, the request logic would re-attempt connection but only once, which instantly failed as well, failing the whole initial api version discovery and disabling all client modules.

While in this case it seems like a problem of test failing to setup Federation before attempting to run client code, it makes me think that in real life there might be cases where the client is joining new Federation, gets a config, and then e.g. switches from wifi to GSM, etc. and has a brief period of being offline.

Again - this is only ever a problem on a first time client is joining. But to prevent such a hiccups, in this change, I'm making it so the client will attempt at least for a certain period to discover the version before giving it up, to lower the chances of such an unfortunate UX.

### CODE REVIEW

**Review by elsirion — APPROVED** [review]

**Comment by elsirion** [comment]
Will this fix  #5286?

**Comment by dpc** [comment]
> Will this fix https://github.com/fedimint/fedimint/issues/5286?

I believe something else already fixed it in @joschisan's PR. But if it had been there, it would probably also prevented it.

**Comment by joschisan** [comment]
@dpc I did not fix this in general, just prevented this by awaiting the api  to come online.

**Comment by dpc** [comment]
> @dpc I did not fix this in general, just prevented this by awaiting the api to come online.

Yeah. Waiting for api to come online was the reason of the problem. The client can't join federation if it's not online yet.

File: fedimint-client/src/lib.rs [hunk 1]
```diff
@@ -1393,6 +1395,33 @@ impl Client {
...
+            if res.is_ok() {
+                return res;
+            }
+
+            if deadline < now() {
+                return res;
+            }
+            debug!(target: LOG_CLIENT, "Retrying getting api version from Federation");
+            sleep(Duration::from_secs(1)).await;
+        }
```
Comment (elsirion) [inline fedimint-client/src/lib.rs:1422]: nit: #5323

**Review by elsirion — APPROVED** [review]

**Pull request merged**